File: 1694-make-sum-divisible-by-p/1694-make-sum-divisible-by-p.py
class Solution:
    def minSubarray(self, nums: List[int], p: int) -> int:
        n = len(nums)
        total = sum(nums)
        if total % p == 0:
            return 0
        if total < p:
            return -1
        target = total % p
        prefix = 0
        prefix_sum = {0: -1}
        min_len = n
        for i in range(n):
            prefix = (prefix + nums[i]) % p
            prefix_sum[prefix] = i
            if (prefix - target) % p in prefix_sum:
                min_len = min(min_len, i - prefix_sum[(prefix - target) % p])
        return min_len if min_len < n else -1
    

File: 1694-make-sum-divisible-by-p/1694-make-sum-divisible-by-p.cpp
class Solution {
public:
    int minSubarray(vector<int>& nums, int p) {
        int n = nums.size();
        long long sum = 0;
        for (int i = 0; i < n; i++) {
            sum += nums[i];
            sum %= p;
        }
        if (sum == 0) {
            return 0;
        }
        unordered_map<long long, int> mp;
        mp[0] = -1;
        long long prefix = 0;
        int ans = n;
        for (int i = 0; i < n; i++) {
            prefix += nums[i];
            prefix %= p;
            long long need = (prefix - sum + p) % p;
            if (mp.find(need) != mp.end()) {
                ans = min(ans, i - mp[need]);
            }
            mp[prefix] = i;
        }
        return ans == n ? -1 : ans;
    }
};